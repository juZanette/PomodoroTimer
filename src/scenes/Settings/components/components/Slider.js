import React from 'react'
import Typography from '@material-ui/core/Typography'
import MatSlider from '@material-ui/core/Slider'
import Box from '@material-ui/core/Box'
import { useDispatch, useSelector } from 'react-redux'

const Slider = ({
  name,
  field,
  defaultValue,
  step,
  marks,
  min,
  max,
  unit,
  action,
  value,
}) => {
  const currentValue = useSelector((state) => state.settings[field])

  const dispatch = useDispatch()

  const handleChange = (value) => {
    if (currentValue !== value) {
      dispatch(action(value))
    }
  }

  const valueText = (value) => {
    return `${value} ${unit}`
  }

  return (
    <Box mt={2}>
      <Typography id={`${name}-slider`} gutterBottom>
        {name}
      </Typography>
      <MatSlider
        defaultValue={defaultValue}
        getAriaValueText={valueText}
        aria-labelledby={`${name}-slider`}
        valueLabelDisplay="auto"
        step={step}
        marks={marks}
        min={min}
        max={max}
        onChangeCommitted={(event, value) => handleChange(value)}
        value={value}
      />
    </Box>
  )
}

export { Slider }
