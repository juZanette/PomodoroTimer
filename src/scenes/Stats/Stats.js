import React from 'react'

const Stats = () => {
  return <div>Stats works</div>
}

export { Stats }
