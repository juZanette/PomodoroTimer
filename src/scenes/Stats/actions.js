export const setRounds = () => ({
  type: 'INCREMENT_TOTAL_SESSIONS',
})
