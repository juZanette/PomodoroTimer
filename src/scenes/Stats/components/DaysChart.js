import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Box from '@material-ui/core/Box'
import MatCard from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import IconButton from '@material-ui/core/IconButton'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import Typography from '@material-ui/core/Typography'
import MatCardContent from '@material-ui/core/CardContent'
import Divider from '@material-ui/core/Divider'
import CardActions from '@material-ui/core/CardActions'
import Chip from '@material-ui/core/Chip'
import { Doughnut } from 'react-chartjs-2'
import dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import { daysOfWeek, colors } from '../data/constants'
import {
  generateChartTitle,
  generateChartTimeLabel,
  generateChartSessionsLabel,
} from '../data/chartFunctions'
import { filters, getFilterFn } from '../data/filters'

dayjs.extend(isSameOrAfter)
require('dayjs/locale/en-gb')
require('dayjs/locale/en')

export const DaysChart = () => {
  const sessions = useSelector((state) => state.sessions)

  const firstDayOfTheWeek = useSelector(
    (state) => state.settings.firstDayOfTheWeek
  )

  const [chartData, setChartData] = useState({
    labels: daysOfWeek,
    datasets: [{ data: Array(7).fill(1), backgroundColor: colors }],
  })

  const theme = useTheme()

  const [chartOptions, setChartOptions] = useState({
    maintainAspectRatio: false,
    legend: {
      labels: {
        fontColor: theme.palette.text.secondary,
      },
    },
    tooltips: {
      callbacks: {
        title: generateChartTitle,
        label: generateChartTimeLabel,
      },
    },
  })

  const [calculatedData, setCalculatedData] = useState()
  const [dataType, setDataType] = useState('time')
  const [anchorEl, setAnchorEl] = useState(null)
  const [filter, setFilter] = useState(filters.find((f) => f.default))

  useEffect(() => {
    if (sessions && sessions.length && firstDayOfTheWeek) {
      let week = [...daysOfWeek]

      if (firstDayOfTheWeek === 'Monday') {
        dayjs.locale('en-gb')
      } else {
        week = daysOfWeek.slice(1, daysOfWeek.length - 1)

        week.unshift('Sunday')
        week.push('Monday')
        dayjs.locale('en')
      }

      const filterFn = getFilterFn(filter)

      const data = Array(7)
        .fill()
        .map(() => ({ sessions: 0, time: 0 }))

      sessions.forEach(({ createdAt, duration }) => {
        if (filterFn(createdAt, dayjs)) {
          const dayIndex = dayjs(createdAt).day()

          data[dayIndex] = {
            sessions: data[dayIndex].sessions + 1,
            time: data[dayIndex].time + getSeconds(duration),
          }
        }
      })

      setCalculatedData(data)

      setChartData((prev) => {
        const newData = { ...prev }
        newData.datasets[0].data = data.map((d) => d[dataType])
        return newData
      })
    }
  }, [sessions, firstDayOfTheWeek, filter, dataType])

  const onChipClicked = (typeSelected) => {
    if (typeSelected === dataType) return

    setDataType(typeSelected)

    setChartData((prev) => {
      const newData = { ...prev }
      newData.datasets[0].data = calculatedData.map(
        (data) => data[typeSelected]
      )
      return newData
    })

    setChartOptions((prev) => {
      const newOptions = { ...prev }

      newOptions.tooltips.callbacks.label =
        typeSelected === 'time'
          ? generateChartTimeLabel
          : generateChartSessionsLabel

      return newOptions
    })
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const closeMenu = () => {
    setAnchorEl(null)
  }

  const handleSelect = (filter) => {
    setFilter(filter)
    closeMenu()
  }

  const getSeconds = ({ minutes, seconds }) => minutes * 60 + seconds

  return (
    <Card theme={theme}>
      <CardHeader
        title="Days of week"
        action={
          <IconButton aria-label="filtering options" onClick={openMenu}>
            <MoreVertIcon />
          </IconButton>
        }
      ></CardHeader>

      <Menu
        id="long-menu"
        anchorEl={anchorEl}
        open={!!anchorEl}
        onClose={closeMenu}
        margin="dense"
      >
        {filters.map((f, i) => (
          <MenuItem
            selected={filter.value === f.value}
            key={i}
            onClick={() => handleSelect(f)}
          >
            {f.name}
          </MenuItem>
        ))}
      </Menu>

      <CardContent>
        <Typography variant="caption" color="textSecondary">
          {(filter.displayName || filter.name || '').toUpperCase()}
        </Typography>

        <ChartContainer>
          <Doughnut data={chartData} options={chartOptions} />
        </ChartContainer>
      </CardContent>

      <Divider variant="middle" />

      <CardActions>
        <div>
          <Box mr={1} display="inline-block">
            <Chip
              onClick={() => onChipClicked('time')}
              label="Time"
              color={dataType === 'time' ? 'secondary' : 'default'}
              clickable
            />
          </Box>

          <Chip
            onClick={() => onChipClicked('sessions')}
            label="Sessions"
            color={dataType === 'sessions' ? 'secondary' : 'default'}
            clickable
          />
        </div>
      </CardActions>
    </Card>
  )
}

const Card = styled(MatCard)`
  margin: 10px auto auto;
  max-width: 350px;

  @media (min-width: 768px) {
    max-width: 450px;
  }

  ${({ theme }) => theme.breakpoints.up('md')} {
    max-width: 550px;
  }
`

const CardContent = styled(MatCardContent)`
  padding-top: 0;
`

const ChartContainer = styled.div`
  height: 250px;
`
