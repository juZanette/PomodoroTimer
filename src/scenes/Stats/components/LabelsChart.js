import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Box from '@material-ui/core/Box'
import MatCard from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import IconButton from '@material-ui/core/IconButton'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import Typography from '@material-ui/core/Typography'
import MatCardContent from '@material-ui/core/CardContent'
import Divider from '@material-ui/core/Divider'
import CardActions from '@material-ui/core/CardActions'
import Chip from '@material-ui/core/Chip'
import { Doughnut } from 'react-chartjs-2'
import { filters, getFilterFn } from '../data/filters'
import {
  generateChartTitle,
  generateChartTimeLabel,
  generateChartSessionsLabel,
} from '../data/chartFunctions'
import * as dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import grey from '@material-ui/core/colors/grey'

dayjs.extend(isSameOrAfter)
require('dayjs/locale/en-gb')
require('dayjs/locale/en')

export const LabelsChart = () => {
  const labels = useSelector((state) => state.labels.data)

  const sessions = useSelector((state) => state.sessions)

  const firstDayOfTheWeek = useSelector(
    (state) => state.settings.firstDayOfTheWeek
  )

  const [chartData, setChartData] = useState({
    labels: [],
    datasets: [{ data: [], backgroundColor: [] }],
  })

  const theme = useTheme()

  const [chartOptions, setChartOptions] = useState({
    maintainAspectRatio: false,
    legend: {
      labels: {
        fontColor: theme.palette.text.secondary,
      },
    },
    tooltips: {
      callbacks: {
        title: generateChartTitle,
        label: generateChartTimeLabel,
      },
    },
  })

  const [calculatedData, setCalculatedData] = useState()
  const [dataType, setDataType] = useState('time')
  const [anchorEl, setAnchorEl] = useState(null)
  const [filter, setFilter] = useState(filters.find((f) => f.default))

  useEffect(() => {
    if (sessions && sessions.length && firstDayOfTheWeek) {
      if (firstDayOfTheWeek === 'Monday') {
        dayjs.locale('en-gb')
      } else {
        dayjs.locale('en')
      }

      const labelsData = labels.map((label) => {
        return {
          ...label,
          sessions: 0,
          time: 0,
        }
      })

      if (!labelsData.length) return

      labelsData.push({
        id: null,
        color: grey[500],
        name: 'Unlabeled',
        sessions: 0,
        time: 0,
      })

      const filterFn = getFilterFn(filter)

      sessions.forEach(({ createdAt, duration, label }) => {
        if (filterFn(createdAt, dayjs)) {
          const index = labelsData.findIndex((l) => l.id === label)

          labelsData[index] = {
            ...labelsData[index],
            sessions: labelsData[index].sessions + 1,
            time: labelsData[index].time + getSeconds(duration),
          }
        }
      })

      setCalculatedData(labelsData)

      setChartData((prev) => {
        const newData = { ...prev }
        newData.datasets[0].data = labelsData.map((d) => d[dataType])
        newData.datasets[0].backgroundColor = labelsData.map((l) => l.color)
        newData.labels = labelsData.map((l) => l.name)

        return newData
      })
    }
  }, [sessions, firstDayOfTheWeek, filter, dataType, labels])

  const onChipClicked = (typeSelected) => {
    if (typeSelected === dataType) return

    setDataType(typeSelected)

    setChartData((prev) => {
      const newData = { ...prev }
      newData.datasets[0].data = calculatedData.map(
        (data) => data[typeSelected]
      )
      return newData
    })

    setChartOptions((prev) => {
      const newOptions = { ...prev }

      newOptions.tooltips.callbacks.label =
        typeSelected === 'time'
          ? generateChartTimeLabel
          : generateChartSessionsLabel

      return newOptions
    })
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const closeMenu = () => {
    setAnchorEl(null)
  }

  const handleSelect = (filter) => {
    setFilter(filter)
    closeMenu()
  }

  const getSeconds = ({ minutes, seconds }) => minutes * 60 + seconds

  return (
    <Card theme={theme}>
      <CardHeader
        title="Labels"
        action={
          <IconButton aria-label="filtering options" onClick={openMenu}>
            <MoreVertIcon />
          </IconButton>
        }
      ></CardHeader>

      <Menu
        id="long-menu"
        anchorEl={anchorEl}
        open={!!anchorEl}
        onClose={closeMenu}
        margin="dense"
      >
        {filters.map((f, i) => (
          <MenuItem
            selected={filter.value === f.value}
            key={i}
            onClick={() => handleSelect(f)}
          >
            {f.name}
          </MenuItem>
        ))}
      </Menu>

      <CardContent>
        <Typography variant="caption" color="textSecondary">
          {(filter.displayName || filter.name || '').toUpperCase()}
        </Typography>

        <ChartContainer>
          <Doughnut data={chartData} options={chartOptions} />
        </ChartContainer>
      </CardContent>

      <Divider variant="middle" />

      <CardActions>
        <div>
          <Box mr={1} display="inline-block">
            <Chip
              onClick={() => onChipClicked('time')}
              label="Time"
              color={dataType === 'time' ? 'secondary' : 'default'}
              clickable
            />
          </Box>

          <Chip
            onClick={() => onChipClicked('sessions')}
            label="Sessions"
            color={dataType === 'sessions' ? 'secondary' : 'default'}
            clickable
          />
        </div>
      </CardActions>
    </Card>
  )
}

const Card = styled(MatCard)`
  margin: 10px auto auto;
  max-width: 350px;

  @media (min-width: 768px) {
    max-width: 450px;
  }

  ${({ theme }) => theme.breakpoints.up('md')} {
    max-width: 550px;
  }
`

const CardContent = styled(MatCardContent)`
  padding-top: 0;
`

const ChartContainer = styled.div`
  height: 250px;
`
