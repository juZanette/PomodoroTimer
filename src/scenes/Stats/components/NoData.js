import React from 'react'
import { Link } from 'react-router-dom'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import MatBox from '@material-ui/core/Box'
import Typography from '@material-ui/core/Typography'
import illustration from './assets/no-data.svg'

export const NoData = () => {
  const theme = useTheme()

  return (
    <Box
      display="flex"
      flexDirection="column"
      justifyContent="center"
      alignItems="center"
      alignSelf="center"
      width="90%"
      maxWidth="600px"
    >
      <img src={illustration} alt="Statistics and charts" />
      <Typography align="center">
        Nothing to show here. Come back after you{' '}
        <LinkStyled to="/timer" color={theme.palette.secondary.main}>
          complete a few sessions
        </LinkStyled>
        .
      </Typography>
    </Box>
  )
}

const Box = styled(MatBox)`
  position: absolute;
  top: 40%;
  left: 50%;
  transform: translate(-50%, -40%);
`

const LinkStyled = styled(Link)`
  text-decoration: none;
  color: ${({ color }) => color};
`
