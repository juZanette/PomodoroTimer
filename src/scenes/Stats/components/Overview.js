import React, { useState, useEffect, useCallback } from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Box from '@material-ui/core/Box'
import MatCard from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import MatCardContent from '@material-ui/core/CardContent'
import Grid from '@material-ui/core/Grid'
import Divider from '@material-ui/core/Divider'
import CardActions from '@material-ui/core/CardActions'
import Chip from '@material-ui/core/Chip'
import dayjs from 'dayjs'
import isToday from 'dayjs/plugin/isToday'
import weekOfYear from 'dayjs/plugin/weekOfYear'
require('dayjs/locale/en-gb')
require('dayjs/locale/en')

dayjs.extend(isToday)
dayjs.extend(weekOfYear)

export const Overview = () => {
  const [dataType, setDataType] = useState('time')
  const [todayStats, setTodayStats] = useState({
    time: null,
    sessions: null,
    avgTime: null,
    avgSessions: null,
  })
  const [weekStats, setWeekStats] = useState({
    time: null,
    sessions: null,
    avgTime: null,
    avgSessions: null,
  })
  const [monthStats, setMonthStats] = useState({
    time: null,
    sessions: null,
    avgTime: null,
    avgSessions: null,
  })

  const [totalStats, setTotalStats] = useState({ time: null, sessions: null })

  const sessions = useSelector((state) => state.sessions)
  const userCreationTime = useSelector((state) => state.auth.creationTime)
  const firstDayOfTheWeek = useSelector(
    (state) => state.settings.firstDayOfTheWeek
  )

  const onChipClicked = (typeSelected) => {
    if (typeSelected === dataType) return

    setDataType(typeSelected)
  }

  const getSeconds = ({ minutes, seconds }) => minutes * 60 + seconds

  const theme = useTheme()

  const calculateAvgTime = useCallback((total, divider) => {
    const d = divider || 1
    const seconds = Math.round((total / d + Number.EPSILON) * 10) / 10
    return calculateTime(seconds)
  }, [])

  useEffect(() => {
    if (sessions && sessions.length && firstDayOfTheWeek) {
      if (firstDayOfTheWeek === 'Monday') {
        dayjs.locale('en-gb')
      } else {
        dayjs.locale('en')
      }

      let today = { time: 0, sessions: 0, avgTime: 0, avgSessions: 0 }
      let week = { time: 0, sessions: 0, avgTime: 0, avgSessions: 0 }
      let month = { time: 0, sessions: 0, avgTime: 0, avgSessions: 0 }
      let total = { time: 0, sessions: 0 }

      sessions.forEach(({ duration, createdAt }) => {
        const secondsCalculated = getSeconds(duration)

        total = updateStats(total, secondsCalculated)

        if (dayjs(createdAt).month() === dayjs().month()) {
          month = updateStats(month, secondsCalculated)

          if (dayjs(createdAt).week() === dayjs().week()) {
            week = updateStats(week, secondsCalculated)

            if (dayjs(createdAt).isToday()) {
              today = updateStats(today, secondsCalculated)
            }
          }
        }
      })

      const now = dayjs()
      const creationTime = dayjs(userCreationTime)
      const daysDiff = now.diff(creationTime, 'd')

      setTodayStats({
        time: calculateTime(today.time),
        sessions: today.sessions,
        avgSessions: calculateAvgSessions(total.sessions, daysDiff),
        avgTime: calculateAvgTime(total.time, daysDiff),
      })

      setWeekStats({
        time: calculateTime(week.time),
        sessions: week.sessions,
        avgSessions: calculateAvgSessions(total.sessions, daysDiff / 7),
        avgTime: calculateAvgTime(total.time, daysDiff / 7),
      })

      setMonthStats({
        time: calculateTime(month.time),
        sessions: month.sessions,
        avgSessions: calculateAvgSessions(total.sessions, daysDiff / 30),
        avgTime: calculateAvgTime(total.time, daysDiff / 30),
      })

      setTotalStats({
        time: calculateTime(total.time),
        sessions: total.sessions,
      })
    }
  }, [calculateAvgTime, firstDayOfTheWeek, sessions, userCreationTime])

  const calculateTime = (time) => {
    const hours = Math.floor(time / 3600)
    const minutes = Math.floor((time % 3600) / 60)

    if (!hours && !minutes) return '-'

    return `${hours}h ${minutes}m`
  }

  const calculateAvgSessions = (total, divider) => {
    if (divider === 0) return total
    return Math.round((total / divider + Number.EPSILON) * 10) / 10
  }

  const updateStats = (period, time) => {
    return {
      time: period.time + time,
      sessions: period.sessions + 1,
    }
  }

  return (
    <Card theme={theme}>
      <CardHeader title="Overview"></CardHeader>
      <CardContent>
        <Grid container spacing={1}>
          <Grid item xs={3}>
            <Box display="flex" flexDirection="column">
              <Sum color={theme.palette.secondary.main}>
                {dataType === 'time'
                  ? todayStats.time || '-'
                  : todayStats.sessions || '-'}
              </Sum>

              <Label>Today</Label>

              <Avg color={theme.palette.text.secondary}>
                {dataType === 'time'
                  ? todayStats.avgTime || '-'
                  : todayStats.avgSessions || '-'}{' '}
                avg
              </Avg>
            </Box>
          </Grid>

          <Grid item xs={3}>
            <Box display="flex" flexDirection="column">
              <Sum color={theme.palette.secondary.main}>
                {dataType === 'time'
                  ? weekStats.time || '-'
                  : weekStats.sessions || '-'}
              </Sum>

              <Label>Week {dayjs().week()}</Label>

              <Avg color={theme.palette.text.secondary}>
                {dataType === 'time'
                  ? weekStats.avgTime || '-'
                  : weekStats.avgSessions || '-'}{' '}
                avg
              </Avg>
            </Box>
          </Grid>

          <Grid item xs={3}>
            <Box display="flex" flexDirection="column">
              <Sum color={theme.palette.secondary.main}>
                {dataType === 'time'
                  ? monthStats.time || '-'
                  : monthStats.sessions || '-'}
              </Sum>

              <Label>{dayjs().format('MMMM')}</Label>

              <Avg color={theme.palette.text.secondary}>
                {dataType === 'time'
                  ? monthStats.avgTime || '-'
                  : monthStats.avgSessions || '-'}{' '}
                avg
              </Avg>
            </Box>
          </Grid>

          <Grid item xs={3}>
            <Box display="flex" flexDirection="column">
              <Sum color={theme.palette.secondary.main}>
                {dataType === 'time'
                  ? totalStats.time || '-'
                  : totalStats.sessions || '-'}
              </Sum>

              <Label>Total</Label>
            </Box>
          </Grid>
        </Grid>
      </CardContent>

      <Divider variant="middle" />

      <CardActions>
        <div>
          <Box mr={1} display="inline-block">
            <Chip
              onClick={() => onChipClicked('time')}
              label="Time"
              color={dataType === 'time' ? 'secondary' : 'default'}
              clickable
            />
          </Box>

          <Chip
            onClick={() => onChipClicked('sessions')}
            label="Sessions"
            color={dataType === 'sessions' ? 'secondary' : 'default'}
            clickable
          />
        </div>
      </CardActions>
    </Card>
  )
}

const Card = styled(MatCard)`
  margin: 10px auto auto;
  max-width: 350px;

  @media (min-width: 768px) {
    max-width: 450px;
  }

  ${({ theme }) => theme.breakpoints.up('md')} {
    max-width: 550px;
  }
`

const CardContent = styled(MatCardContent)`
  padding-top: 0;
`

const Sum = styled.span`
  display: inline-block;
  text-align: center;
  color: ${({ color }) => color};
  font-weight: bold;
`

const Label = styled.span`
  display: inline-block;
  text-align: center;
  margin: 2px 0;
`

const Avg = styled.span`
  display: inline-block;
  text-align: center;
  font-size: 0.7rem;
  color: ${({ color }) => color};
`

export { Sum, Avg }
