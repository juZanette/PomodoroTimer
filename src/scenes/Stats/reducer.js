const initialState = {
  totalSessions: 0,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'INCREMENT_TOTAL_SESSIONS':
      return {
        totalSessions: ++state.totalSessions,
      }
    default:
      return state
  }
}
