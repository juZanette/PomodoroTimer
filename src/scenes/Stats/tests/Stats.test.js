import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { createMount } from '@material-ui/core/test-utils'
import { Stats } from '../Stats'
import sessions from '../../../data/sessions/tests/mock-data/sessions'
import labels from '../../../data/labels/tests/mock-data/labels'
import { BrowserRouter } from 'react-router-dom'
import { NoData } from '../components/NoData'

jest.mock('react-chartjs-2', () => ({
  Doughnut: () => null,
}))

describe('<Stats />', () => {
  const mockStore = configureMockStore([thunk])
  let store
  let mount
  let wrapper

  beforeEach(() => {
    mount = createMount()

    wrapper = (sessions = []) => {
      store = mockStore({
        auth: {
          uid: 's3sns123',
          creationTime: new Date('05/01/2020'),
          name: null,
          photo: null,
        },
        settings: {
          firstDayOfTheWeek: 'Monday',
        },
        labels: {
          data: labels,
        },
        sessions,
      })

      store.dispatch = jest.fn()

      return mount(
        <Provider store={store}>
          <BrowserRouter>
            <Stats />
          </BrowserRouter>
        </Provider>
      )
    }
  })

  afterEach(() => {
    mount.cleanUp()
  })

  test('should render <Stats /> correctly', () => {
    expect(wrapper()).toMatchSnapshot()
  })

  test('should render <NoData /> if no sessions are provided', () => {
    expect(wrapper().find(NoData).exists()).toBe(true)
  })

  test('should not render <NoData /> if the sessions are provided', () => {
    expect(wrapper(sessions).find(NoData).exists()).toBe(false)
  })
})
