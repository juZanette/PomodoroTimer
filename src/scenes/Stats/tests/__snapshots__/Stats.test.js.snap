// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stats /> should render <Stats /> correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Stats>
        <div>
          <NoData>
            <Styled(Styled(MuiBox))
              alignItems="center"
              alignSelf="center"
              display="flex"
              flexDirection="column"
              justifyContent="center"
              maxWidth="600px"
              width="90%"
            >
              <Styled(MuiBox)
                alignItems="center"
                alignSelf="center"
                className="sc-fzpans itosWR"
                display="flex"
                flexDirection="column"
                justifyContent="center"
                maxWidth="600px"
                width="90%"
              >
                <div
                  className="MuiBox-root MuiBox-root-2 sc-fzpans itosWR"
                >
                  <img
                    alt="Statistics and charts"
                    src="no-data.svg"
                  />
                  <WithStyles(ForwardRef(Typography))
                    align="center"
                  >
                    <ForwardRef(Typography)
                      align="center"
                      classes={
                        Object {
                          "alignCenter": "MuiTypography-alignCenter",
                          "alignJustify": "MuiTypography-alignJustify",
                          "alignLeft": "MuiTypography-alignLeft",
                          "alignRight": "MuiTypography-alignRight",
                          "body1": "MuiTypography-body1",
                          "body2": "MuiTypography-body2",
                          "button": "MuiTypography-button",
                          "caption": "MuiTypography-caption",
                          "colorError": "MuiTypography-colorError",
                          "colorInherit": "MuiTypography-colorInherit",
                          "colorPrimary": "MuiTypography-colorPrimary",
                          "colorSecondary": "MuiTypography-colorSecondary",
                          "colorTextPrimary": "MuiTypography-colorTextPrimary",
                          "colorTextSecondary": "MuiTypography-colorTextSecondary",
                          "displayBlock": "MuiTypography-displayBlock",
                          "displayInline": "MuiTypography-displayInline",
                          "gutterBottom": "MuiTypography-gutterBottom",
                          "h1": "MuiTypography-h1",
                          "h2": "MuiTypography-h2",
                          "h3": "MuiTypography-h3",
                          "h4": "MuiTypography-h4",
                          "h5": "MuiTypography-h5",
                          "h6": "MuiTypography-h6",
                          "noWrap": "MuiTypography-noWrap",
                          "overline": "MuiTypography-overline",
                          "paragraph": "MuiTypography-paragraph",
                          "root": "MuiTypography-root",
                          "srOnly": "MuiTypography-srOnly",
                          "subtitle1": "MuiTypography-subtitle1",
                          "subtitle2": "MuiTypography-subtitle2",
                        }
                      }
                    >
                      <p
                        className="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
                      >
                        Nothing to show here. Come back after you
                         
                        <Styled(Link)
                          color="#f50057"
                          to="/timer"
                        >
                          <Link
                            className="sc-fzplWN cOTjdd"
                            color="#f50057"
                            to="/timer"
                          >
                            <LinkAnchor
                              className="sc-fzplWN cOTjdd"
                              color="#f50057"
                              href="/timer"
                              navigate={[Function]}
                            >
                              <a
                                className="sc-fzplWN cOTjdd"
                                color="#f50057"
                                href="/timer"
                                onClick={[Function]}
                              >
                                complete a few sessions
                              </a>
                            </LinkAnchor>
                          </Link>
                        </Styled(Link)>
                        .
                      </p>
                    </ForwardRef(Typography)>
                  </WithStyles(ForwardRef(Typography))>
                </div>
              </Styled(MuiBox)>
            </Styled(Styled(MuiBox))>
          </NoData>
        </div>
      </Stats>
    </Router>
  </BrowserRouter>
</Provider>
`;
